scripts/bin/update-packages

```diff
@@ -148,22 +151,130 @@ _fetch_and_cache_tags() {
+		# Newlines need to be escaped for the query parser
+		GITHUB_GRAPHQL_QUERIES+=( "_$((COUNTER++)): repository(owner: \\\"${OWNER}\\\", name: \\\"${REPO}\\\") { ..._${TAG_TYPE//-/_} }" )
+
+		unset PKG_SRCURL TAG_TYPE OWNER REPO
+	done
+
+	# Batch size for fetching tags, 100 seems to work consistently.
+	local BATCH BATCH_SIZE=100
+	echo "// Begin test: $(date)" > /tmp/query-12345
+	for (( BATCH = 0; ${#GITHUB_GRAPHQL_QUERIES[@]} >= BATCH_SIZE * BATCH ; BATCH++ )); do
+
+		echo "Starting batch $BATCH at: ${GITHUB_GRAPHQL_QUERIES[$BATCH * $BATCH_SIZE]//\\/}"
+
+		# JSON strings cannot contain tabs or newlines
+		# so shutup shellcheck complaining about escapes in single quotes
+		local QUERY
+
+		# Start the GraphQL query with our two fragments for getting the latest tag from a release, and from refs/tags
+		# These are defined only if needed, so this one is for '_latest_release_tag'
+		grep -q '_latest_release_tag' <<< "${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" && {
+			QUERY+="$(printf '%s\n' \
+			'fragment _latest_release_tag on Repository {' \
+			'  latestRelease { tagName }' \
+			'}')"
+		}
+
+		grep -q '_newest_tag' <<< "${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" && {
+			QUERY+="$(printf '%s\n' \
+			'fragment _newest_tag on Repository {' \
+			'  refs( refPrefix: \"refs/tags/\" first: 1 orderBy: {field: TAG_COMMIT_DATE, direction: DESC}) {' \
+			'    nodes { name }' \
+			'  }' \
+			'}')"
+		}
+
+		QUERY+='query {'
+
+		# Fill out the query body with the package repos we need to query for updates
+		# Lastly fetch the rate limit utilization
+		echo "// Batch: $BATCH" >> /tmp/query-12345
+		printf -v QUERY '%s\n' \
+				"${QUERY}" \
+				"${GITHUB_GRAPHQL_QUERIES[@]:$BATCH * $BATCH_SIZE:$BATCH_SIZE}" \
+				'ratelimit: rateLimit { cost limit remaining used resetAt }' \
+				'}' \
+
+		printf '%s' "${QUERY}" >> /tmp/query-12345
+		response="$(printf '{ "query": "%s" }' "${QUERY//$'\n'/ }" | curl -fsL \
+			-H "Authorization: token ${GITHUB_TOKEN}" \
+			-H "Accept: application/vnd.github.v3+json" \
+			-H 'Content-Type: application/json' \
+			-X POST \
+			--data @- \
+			https://api.github.com/graphql
+		)"
+		unset QUERY
+		local TAGS idx i=0
+		TAGS="$(jq -r '.data            # From the data: table
+			| del(.ratelimit)           # Remove the ratelimit: table
+			| to_entries[]              # convert the remaining entries to an array
+			| .value                    # For each .value
+			| (.latestRelease?.tagName  # Print out the tag name of the latest release
+			// .refs.nodes[]?.name      # or of the latest tag
+			// null)' <<< "$response")" # If neither exists return null
+
+		while IFS=$'\n' read -r idx; do
+			printf 'PKG|%s|%s\n' \
+				"${__GITHUB_PACKAGES[$BATCH * $BATCH_SIZE + $(( i++ ))]##*/}" \
+				"${idx}"
+		done <<< "$TAGS"
+	done
+
+	# This is called indirectly in the ninja file
+	# So silence shellcheck's unreachable code warning
+	# shellcheck disable=SC2317
```

Comment:
We can go a termux_download way and move the whole fetch-and-cache shenanigans to separate bash script which can be both sourced an executed.
https://github.com/termux/termux-packages/blob/25ef2aaa5cc4c0da636b04f82fa17fee2e3b37ea/scripts/build/termux_download.sh#L70-L73

In this case this solution can be avoided and will make the code a bit more well-organized.

Reply:
That was the plan.